src/libcore/iter.rs

```diff
@@ -1791,6 +1797,16 @@ impl<T, I> Iterator for Peekable<T, I> where I: Iterator<Item=T> {
 }
 
 #[stable]
+impl<T, I> ExactSizeIterator for Peekable<T, I> where I: ExactSizeIterator<Item = T> {
+    #[inline]
+    fn len(&self) -> usize {
+        // This is guarenteed to not overflow because `len()` must have been able to return a valid
+        // value before we peeked.
+        self.iter.len() + if self.peeked.is_some() { 1 } else { 0 }
+    }
+}
```

Comment:
I don't think we can quite add this implementation just yet as I believe it's possible to return `(usize::MAX, None)` from this iterator (which isn't an exact size).